bids/layout/models.py

```diff
@@ -294,20 +290,34 @@ def get_df(self, include_timing=True, adjust_onset=False):
                 timeseries file is shifted to reflect the "StartTime" value in
                 the JSON sidecar. If False, the first sample starts at 0 secs.
                 Ignored if include_timing=False.
+            enforce_dtypes (bool): If True, enforces the data types defined in
+                the BIDS spec on core columns (e.g., subject_id and session_id
+                must be represented as strings).
 
         Returns: A pandas DataFrame.
         """
         import pandas as pd
         import numpy as np
 
-        # TODO: memoize method instead of just caching the raw data
-        if self._data is None:
-            self._data = pd.read_csv(self.path, sep='\t', na_values='n/a')
+        if enforce_dtypes:
+            dtype = {
+                'subject_id': str,
+                'session_id': str,
+                'participant_id': str
+            }
+        else:
+            dtype = None
+
+        # TODO: memoize this for efficiency. (Note: caching is insufficient,
+        # because the dtype enforcement will break if we ignore the value of
+        # enforce_dtypes).
+        self.data = pd.read_csv(self.path, sep='\t', na_values='n/a',
+                                    dtype=dtype)
```

Comment:
Yeah I made a note of that already in the above (resolved) threads. Memoizing `self.data` instead of caching was already a TODO, but I had to remove the caching because the suggestions @yarikoptic made (which I agree were necessary). Still worth doing, but not a terribly high priority, and it should probably be a separate PR with a more general solution (i.e., at `BIDSFile` level rather than `BIDSDataFile`, so that classes like `BIDSJSONFile` can also benefit).